packages/runtimes/js/src/JSRuntime.js

```diff
@@ -67,135 +67,80 @@ export default new Runtime({
       return;
     }
 
-    let loaders = getLoaders(bundle.env);
+    let asyncDependencies = [];
+    let otherDependencies = [];
 
-    // Determine if we need to add a dynamic import() polyfill, or if all target browsers support it natively.
-    let needsDynamicImportPolyfill = false;
-    if (bundle.env.isBrowser() && bundle.env.outputFormat === 'esmodule') {
-      needsDynamicImportPolyfill = !bundle.env.matchesEngines(
-        DYNAMIC_IMPORT_BROWSERS,
-      );
+    for (let dependency of bundleGraph.getDependenciesInBundle(bundle)) {
+      if (dependency.isAsync && !dependency.isURL) {
+        asyncDependencies.push(dependency);
+      } else {
+        otherDependencies.push(dependency);
+      }
     }
 
     let assets = [];
-    for (let dependency of bundleGraph.getExternalDependencies(bundle)) {
-      let bundleGroup = bundleGraph.resolveExternalDependency(dependency);
-      if (bundleGroup == null) {
-        if (dependency.isURL) {
-          // If a URL dependency was not able to be resolved, add a runtime that
-          // exports the original moduleSpecifier.
-          assets.push({
-            filePath: __filename,
-            code: `module.exports = ${JSON.stringify(
-              dependency.moduleSpecifier,
-            )}`,
-            dependency,
-          });
-        }
+    for (let dependency of asyncDependencies) {
+      let resolved = bundleGraph.resolveExternalDependency(bundle, dependency);
+      if (resolved == null) {
         continue;
       }
 
-      let bundlesInGroup = bundleGraph.getBundlesInBundleGroup(bundleGroup);
-
-      let [firstBundle] = bundlesInGroup;
-      if (firstBundle.isInline) {
+      if (resolved.type === 'asset') {
+        // If this bundle already has the asset this dependency references,
+        // return a simple runtime of `Promise.resolve(parcelRequire("assetId"))`.
         assets.push({
-          filePath: path.join(__dirname, `/bundles/${firstBundle.id}.js`),
-          code: `module.exports = ${JSON.stringify(dependency.id)};`,
+          filePath: __filename,
+          code: `module.exports = Promise.resolve(parcelRequire(${JSON.stringify(
```

Comment:
I'm not sure it's always safe to use `parcelRequire` here - that relies on an implementation detail of the packager. For example, the commonjs and ESM output formats don't have this. We'll need to make a real dependency with `require` which will end up getting compiled correctly by the packager.

Reply:
Yeah, this is what's breaking the scope hoisting tests.

How can we create a real dependency on the asset from here? `require()`? If so, what would we use as the module specifier?